fairseq2n/src/fairseq2n/data/sample_data_source.h

```diff
@@ -0,0 +1,51 @@
+// Copyright (c) Meta Platforms, Inc. and affiliates.
+// All rights reserved.
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+
+#pragma once
+
+#include <memory>
+#include <vector>
+
+#include <ATen/Generator.h>
+#include <ATen/Tensor.h>
+
+#include "fairseq2n/data/data_pipeline.h"
+#include "fairseq2n/data/data_source.h"
+
+namespace fairseq2n::detail {
+
+/// @brief sample from a list of datasources
+class sample_data_source final : public data_source {
+public:
+    explicit
+    sample_data_source(std::vector<data_pipeline> &&pipelines,
+     std::vector<float> &&weights);
```

Comment:
Nit: no need for a line break here. You can put this parameter next to `pipelines`.